### DISCUSSION TITLE
How not to refresh the page when using revalidatePath

### DISCUSSION THREAD

**Discussion by hnykda** [vercel/next.js#61836] [Help]
### Summary

Hi,

I have this minimal example: https://gist.github.com/hnykda/835b0daff2410a92f14e6a14bb84a535 . A simple app that fetches 5 wine names from an API on the server (as recommended by the latest nextjs docs), sort them by length, then pass the promise to the client component that use the `use` hook, then display these strings and gives an option to move them up and down. This order should be preserved, that's simulated by the `moveItemSideEffect` - imagine that it e.g. changes an order number in the wine database. To make sure that this change would be visible when the user e.g. visit the page again or open it in a new tab, I have to call `revalidatePath('/')` in the server action (the sideEffect).

The problem I have is that it refreshes the current page, fetching the data again, refreshing everything and losing the current state then, I made a recording here.

Is there a solution to this, or is there something I am doing wrong? Should I just fetch the data on the client so I don't have to deal with the server-sent data?

### Example

https://gist.github.com/hnykda/835b0daff2410a92f14e6a14bb84a535

**Comment by DevDaniloFerrari** [comment]
Your implementation is correct.
Assuming that by changing the position of the wine you would make this change in the database. Then, the api would return the array in the way you organized it on the screen.

**Reply by hnykda** [reply]
Thank you - but it doesn't, as I sort by length in the data fetch. It's understandably a stupid example, as I am controlling the very same thing I am overriding. Stay with me while I complicate the example just a little bit by adding one extra state variable that holds a "focused" element:

![image](https://github.com/vercel/next.js/assets/2741256/d6cb144b-69c9-49ec-9047-af62823cef93)

the code:

```
"use client";
import React, { useState, useEffect } from "react";

const Comp = ({
  existingDataPromise,
  moveItemSideEffect,
}: {
  existingDataPromise: Promise<string[]>;
  moveItemSideEffect: () => void;
}) => {
  const [dataList, setDataList] = useState<string[]>([]);
  const [focused, setFocused] = useState<string>("");

  useEffect(() => {
    existingDataPromise.then((data) => {
      setDataList(data);
      setFocused(data[0]);
    });
  }, [existingDataPromise]);

  const moveItem = (index: number, direction: "up" | "down") => {
    setDataList((prevList) => {
      const newList = [...prevList];
      const element = newList.splice(index, 1)[0];
      const newIndex = direction === "up" ? index - 1 : index + 1;
      newList.splice(newIndex, 0, element);
      return newList;
    });
    moveItemSideEffect();
  };

  return (
    <div className="flex flex-col">
      {dataList.map((item, index) => (
        <div key={index} className="flex flex-row gap-2">
          {item}
          <button
            onClick={() => moveItem(index, "up")}
            disabled={index === 0}
            className="bg-red-500"
          >
            Up
          </button>
          <button
            onClick={() => moveItem(index, "down")}
            disabled={index === dataList.length - 1}
            className="bg-green-500"
          >
            Down
          </button>
          <button
            onClick={() => setFocused(item)}
            className="bg-blue-500"
          >
            Make focused
          </button>
        </div>
      ))}
      <div className="text-xl mt-5">
        Focused: <span className="underline">{focused}</span>
      </div>
    </div>
  );
};

export default Comp;
```

I don't want to store that `focused` element in any way - it's just a client only thing.


...now when I am writing this, I realized that the page is _not_ refreshed, just new data came in and it's on me to reconsolidate the state, e.g. by adding a condition in the `useEffect` to just not mess with it if it has been populated already:

```
 useEffect(() => {
    existingDataPromise.then((data) => {
      if (dataList.length === 0) {
        setDataList(data);
      }
      if (!focused) {
        setFocused(data[0]);
      }
    });
  }, [existingDataPromise, focused, dataList]);
```
